### DISCUSSION TITLE
Scope of package and questions

### DISCUSSION THREAD

**Discussion by AriMKatz** [quinnj/Interfaces.jl#20] [Q&A] [answered]
Hi!

I just ran into this package. 

How does it differentiate itself from various other trait-like packages such as https://github.com/mauro3/SimpleTraits.jl and https://github.com/andyferris/Traitor.jl and https://github.com/tk3369/BinaryTraits.jl?

specifically with the latter, you can do a rudimentary form of typechecking for implemented interface methods. 

Is interfaces meant not to affect dispatch? ie can I dispatch on the presence of an interfaces? What about a generic vs refined interfaces, such as a fastpath for an o(1) indexed datastructure? 

Thanks!

**Comment by nickrobinson251** [comment, marked as answer]
Hello! 

I'm sure Jacob could give a better answer this is primarily Jacob's idea, but i'll give it a go 😊 

> I just ran into this package

cool! out of interest, how'd you come across it? it's sufficiently WIP that there's no README (we will fix that soon!), so kinda cool you came across it already :) For now, the code in the tests is probably the best intro we have right now.

> How does it differentiate itself from various other trait-like packages

I suppose this is kinda related to traits... but i hadn't really put that together! I think it's related to traits (in the "Holy traits" sense) only in so far as many interfaces include defining traits (like the "AbstractArray" interface has `IndexStyle`, for example). But interfaces don't _have_ to have traits, they can just have required methods (which might have required return types but might not, and don't have to return things you can dispatch on). I'm not familair with any of the packages you linked, so can't personally comment on specific differences (but a quick look at the READMEs suggests they are trying to solve a different problem to Interfaces.jl)  

Interfaces.jl is trying to provide the answer to two questions: 
1. "As a developer of a SomePackage (e.g. Base, Tables.jl, ...), which defines some interface (e.g. `AbstractArray`, "Table"), how can I tell users how to create their own types which fulfil this interface (given that e.g. `MyArray <: AbstractArray` isn't the full story, and `MyTable <: Table` isn't any part of the story)?"
2. "As a developer of SomeType (e.g. `Array`, `DataFrame`), which intends to implement some interface (e.g. `AbstractArray`, "Table"), how can I be sure i've actually implemented the interface correctly?"

I think the current answers are
1. Write docs! And consider defining a test suite that all AbstractArray/Table/etc should pass (e.g. FilePathsBase does this)
2. Read docs! And run a test suite if provided by the interface, and/or try out your AbstractArray/Table/whatever in many places you expect an AbstractArray/Table/whatever to work and see what happens

Also sometimes there's no well define answer to "(1) what is the interface of `X`? And (2) how can I be sure `MyType` satsifies the interface?" (I think `AbstractDict` is such an example: https://github.com/JuliaLang/julia/issues/25941). And while Interfaces.jl can't itself provide answers where there isn't yet one, it can provide a _programmatic_ way to declare an interface... which _may_ lead to clearer definitions of what is and is not part of the interface (but that's my speculation, and also i'm not saying that this would be the right tool for all occasions, etc) 
 
> Is interfaces meant not to affect dispatch? ie can I dispatch on the presence of an interfaces? 

I don't think interfaces will affect this. If there is e.g. an `AbstractArray` interface, then you can know that your function
`foo(x::AbstractArray) = ...` is safe to assume about its input `x` e.g. that `size(x)` is defined. Similarly, you can know that your `BarArray <: AbstractArray` will work anywhere that `AbstractArray`s work (so long as `@implements BarArray AbstractArray`).
But if there's e.g. a "Tables" interface, which _doesn't_ involve some common supertype (e.g.`AbstractTable`), then there would be nothing to dispatch on, and you'd just define `foo(table) = ...`, and in the body use methods that are safe to assume exist for a "Table". You _could_ do `foo(table) = (@assert implements(T, TableInterface); ...)`, but usually that's probably not going to be any
benefit over leaving the assert out and just hitting some error (usually a MethodError) soon after if `table` doesn't have the reqired methods (i.e. it's not a "Table").

> What about a generic vs refined interfaces, such as a fastpath for an o(1) indexed datastructure?

I'm not familiar with the terminology "generic" and "refined" interfaces, sorry. I know there are many notations of "interfaces", and likely the terminology differs across programming languages too, but i'm not up to speed with that either, so hoepfully someone else can give a better answer here. But all that said, Interfaces.jl intends to only provide a way to declare an interface in terms of things that can be known statically (e.g. "there must be a `IteratorSize(iter)` method and it must return an object of type `HasLength`, `HasShape`, `IsInfinite`, or `SizeUnknown`") and not in terms of things that can only be known at run-time (e.g. "there must be a `predict(x)` method and it must return an object `y` with `size(x) == size(y)`). So, the `AbstractDict` interface definition won't be able to state "`getindex` must be O(1)".

I think the Interfaces.jl meaning of "interface" aligns closely with the usage on the "interfaces" page of the Julia docs: https://docs.julialang.org/en/v1/manual/interfaces

**Comment by quinnj** [comment]
Yeah, @nickrobinson251 hit the nail on the head here. This is different from other packages in that the primary motivation _isn't_ dispatch-related.
